src/main/scala/is/hail/expr/FunctionRegistry.scala

```diff
@@ -1890,6 +1890,15 @@ object FunctionRegistry {
     "Compute the sum by index. All elements in the aggregable must have the same length."
   )(aggregableHr(arrayHr(doubleHr)), arrayHr(doubleHr))
 
+  val productAggDocstring = """Compute the product of all non-missing elements. The empty product is one."""
+  registerAggregator[Int, Long]("product", () => new IntProductAggregator(), productAggDocstring)(aggregableHr(intHr), longHr)
```

Comment:
You should just register one for longs and one for doubles. The implicit conversions will do the rest. This also means we don't need the IntProductAgg and FloatProductAgg stuff.